File: apps/desktop/src/lib/tilemap.ts
// Bitwright - Sprite Engine
// Copyright (C) 2026 Afterhours Studio
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

/**
 * The tilemap commands, and the event a map write emits.
 *
 * One function per command in `commands/tilemap.rs`, named after it. Every one
 * returns a {@link ShellResult}, like the rest of the bridge in `lib/tauri.ts`,
 * and the argument names are camelCase because Tauri 2 renames them.
 */

import type { UnlistenFn } from '@tauri-apps/api/event';

import { invoke, on } from '@/lib/tauri';
import type { ShellResult } from '@/lib/tauri';
import type { Placement, TileAsset, Tilemap, TilemapChangedEvent } from '@/types/tilemap';

/** The channel `commands/tilemap.rs` and the MCP host emit a map write on. */
export const TILEMAP_CHANGED = 'tilemap://changed';

/** Gives a background asset an empty map with one `ground` layer; refused if it has one. */
export function tilemapCreate(
  assetId: string,
  tileWidth: number,
  tileHeight: number,
  columns: number,
  rows: number,
): Promise<ShellResult<Tilemap>> {
  return invoke<Tilemap>('tilemap_create', { assetId, tileWidth, tileHeight, columns, rows });
}

/** Reads a background's map; `tilemap.none` when it has none yet. */
export function tilemapRead(assetId: string): Promise<ShellResult<Tilemap>> {
  return invoke<Tilemap>('tilemap_read', { assetId });
}

/** Sets cells on one layer, all or none; answers with the map as written. */
export function tilemapPlace(
  assetId: string,
  layer: string,
  placements: Placement[],
): Promise<ShellResult<Tilemap>> {
  return invoke<Tilemap>('tilemap_place', { assetId, layer, placements });
}

/** Adds an empty layer in front of the others. */
export function tilemapAddLayer(
  assetId: string,
  name: string,
  parallax: number,
): Promise<ShellResult<Tilemap>> {
  return invoke<Tilemap>('tilemap_add_layer', { assetId, name, parallax });
}

/** Removes a layer; the last one is refused. */
export function tilemapRemoveLayer(assetId: string, name: string): Promise<ShellResult<Tilemap>> {
  return invoke<Tilemap>('tilemap_remove_layer', { assetId, name });
}

/** Changes a layer's parallax, its visibility, or both. */
export function tilemapSetLayer(
  assetId: string,
  name: string,
  parallax?: number,
  visible?: boolean,
): Promise<ShellResult<Tilemap>> {
  return invoke<Tilemap>('tilemap_set_layer', { assetId, name, parallax, visible });
}

/** The project's tile assets that fit the map, each with a preview, for the picker. */
export function tilemapTiles(assetId: string): Promise<ShellResult<TileAsset[]>> {
  return invoke<TileAsset[]>('tilemap_tiles', { assetId });
}

/** The map rendered, or one layer of it, as a PNG data URL. */
export function tilemapPreview(assetId: string, layer?: string): Promise<ShellResult<string>> {
  return invoke<string>('tilemap_preview', { assetId, layer });
}

/** Subscribes to map writes, whether made here or by an agent. */
export function onTilemapChanged(
  handler: (event: TilemapChangedEvent) => void,
): Promise<UnlistenFn> {
  return on<TilemapChangedEvent>(TILEMAP_CHANGED, handler);
}

File: apps/desktop/src-tauri/src/commands/tilemap.rs
// Bitwright - Sprite Engine
// Copyright (C) 2026 Afterhours Studio
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! Tauri commands exposing the tilemap model (`raster::tilemap`) and its
//! storage (`Store::tilemap_*`, `asset_composite`) to the window.

use crate::commands::document::{run, DocumentState};
use crate::raster::{self, Placement, Tilemap};
use crate::store::{AppError, AssetId, Result, Store};
use serde::Serialize;
use tauri::{AppHandle, Emitter, Runtime, State};

pub const EVENT_TILEMAP_CHANGED: &str = "tilemap://changed";

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TilemapChanged {
    pub asset_id: AssetId,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TileAsset {
    pub id: AssetId,
    pub name: String,
    pub width: u16,
    pub height: u16,
    pub preview: String,
}

fn emit_changed<R: Runtime>(app: &AppHandle<R>, asset_id: AssetId) {
    if let Err(error) = app.emit(EVENT_TILEMAP_CHANGED, TilemapChanged { asset_id }) {
        log::warn!("{EVENT_TILEMAP_CHANGED} failed after the tilemap committed: {error}");
    }
}

fn create(
    store: &mut Store,
    asset_id: AssetId,
    tile_width: u16,
    tile_height: u16,
    columns: u16,
    rows: u16,
) -> Result<Tilemap> {
    if store.tilemap_read(asset_id).is_ok() {
        return Err(AppError::new(
            "tilemap.exists",
            "a tilemap already exists for this asset",
        ));
    }
    let map = Tilemap::new(tile_width, tile_height, columns, rows)?;
    store.tilemap_write(asset_id, &map)
}

fn place(
    store: &mut Store,
    asset_id: AssetId,
    layer: String,
    placements: Vec<Placement>,
) -> Result<Tilemap> {
    let mut map = store.tilemap_read(asset_id)?;
    map.place(&layer, &placements)?;
    store.tilemap_write(asset_id, &map)
}

fn add_layer(store: &mut Store, asset_id: AssetId, name: String, parallax: f32) -> Result<Tilemap> {
    let mut map = store.tilemap_read(asset_id)?;
    map.add_layer(&name, parallax)?;
    store.tilemap_write(asset_id, &map)
}

fn remove_layer(store: &mut Store, asset_id: AssetId, name: String) -> Result<Tilemap> {
    let mut map = store.tilemap_read(asset_id)?;
    map.remove_layer(&name)?;
    store.tilemap_write(asset_id, &map)
}

fn set_layer(
    store: &mut Store,
    asset_id: AssetId,
    name: String,
    parallax: Option<f32>,
    visible: Option<bool>,
) -> Result<Tilemap> {
    let mut map = store.tilemap_read(asset_id)?;
    map.set_layer(&name, parallax, visible)?;
    store.tilemap_write(asset_id, &map)
}

fn tiles(store: &mut Store, asset_id: AssetId) -> Result<Vec<TileAsset>> {
    let map = store.tilemap_read(asset_id)?;
    let asset = store.asset_read(asset_id)?;
    let assets = store.asset_list(asset.project_id)?;
    assets
        .into_iter()
        .filter(|a| a.kind == "tile" && a.width == map.tile_width && a.height == map.tile_height)
        .map(|a| {
            let composite = store.asset_composite(a.id)?;
            let preview = format!(
                "data:image/png;base64,{}",
                raster::png::encode_base64(&composite)?
            );
            Ok(TileAsset {
                id: a.id,
                name: a.name,
                width: a.width,
                height: a.height,
                preview,
            })
        })
        .collect()
}

fn preview(store: &mut Store, asset_id: AssetId, layer: Option<String>) -> Result<String> {
    let image = store.tilemap_render(asset_id, layer.as_deref())?;
    Ok(format!(
        "data:image/png;base64,{}",
        raster::png::encode_base64(&image)?
    ))
}

#[tauri::command]
pub async fn tilemap_create<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DocumentState>,
    asset_id: AssetId,
    tile_width: u16,
    tile_height: u16,
    columns: u16,
    rows: u16,
) -> Result<Tilemap> {
    let map = run(&state, move |s| {
        create(s, asset_id, tile_width, tile_height, columns, rows)
    })
    .await?;
    emit_changed(&app, asset_id);
    Ok(map)
}

#[tauri::command]
pub async fn tilemap_read(state: State<'_, DocumentState>, asset_id: AssetId) -> Result<Tilemap> {
    run(&state, move |s| s.tilemap_read(asset_id)).await
}

#[tauri::command]
pub async fn tilemap_place<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DocumentState>,
    asset_id: AssetId,
    layer: String,
    placements: Vec<Placement>,
) -> Result<Tilemap> {
    let map = run(&state, move |s| place(s, asset_id, layer, placements)).await?;
    emit_changed(&app, asset_id);
    Ok(map)
}

#[tauri::command]
pub async fn tilemap_add_layer<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DocumentState>,
    asset_id: AssetId,
    name: String,
    parallax: f32,
) -> Result<Tilemap> {
    let map = run(&state, move |s| add_layer(s, asset_id, name, parallax)).await?;
    emit_changed(&app, asset_id);
    Ok(map)
}

#[tauri::command]
pub async fn tilemap_remove_layer<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DocumentState>,
    asset_id: AssetId,
    name: String,
) -> Result<Tilemap> {
    let map = run(&state, move |s| remove_layer(s, asset_id, name)).await?;
    emit_changed(&app, asset_id);
    Ok(map)
}

#[tauri::command]
pub async fn tilemap_set_layer<R: Runtime>(
    app: AppHandle<R>,
    state: State<'_, DocumentState>,
    asset_id: AssetId,
    name: String,
    parallax: Option<f32>,
    visible: Option<bool>,
) -> Result<Tilemap> {
    let map = run(&state, move |s| {
        set_layer(s, asset_id, name, parallax, visible)
    })
    .await?;
    emit_changed(&app, asset_id);
    Ok(map)
}

#[tauri::command]
pub async fn tilemap_tiles(
    state: State<'_, DocumentState>,
    asset_id: AssetId,
) -> Result<Vec<TileAsset>> {
    run(&state, move |s| tiles(s, asset_id)).await
}

#[tauri::command]
pub async fn tilemap_preview(
    state: State<'_, DocumentState>,
    asset_id: AssetId,
    layer: Option<String>,
) -> Result<String> {
    run(&state, move |s| preview(s, asset_id, layer)).await
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::raster::Placement;
    use crate::store::Store;

    fn background(store: &mut Store) -> AssetId {
        let project = store.project_create("Project", "gameboy").unwrap();
        let asset = store
            .asset_create(project.id, "Map", "background", 1, 1)
            .unwrap();
        asset.id
    }

    fn tile(store: &mut Store, project_id: uuid::Uuid, w: u16, h: u16) -> AssetId {
        // Asset names are unique within a project, so each tile gets its own.
        let name = format!("Tile {}", uuid::Uuid::now_v7());
        let asset = store.asset_create(project_id, &name, "tile", w, h).unwrap();
        asset.id
    }

    #[test]
    fn create_twice_is_exists() {
        let mut store = Store::memory().unwrap();
        let asset_id = background(&mut store);
        create(&mut store, asset_id, 16, 16, 4, 4).unwrap();
        let err = create(&mut store, asset_id, 16, 16, 4, 4).unwrap_err();
        assert_eq!(err.code, "tilemap.exists");
    }

    #[test]
    fn place_then_read_back() {
        let mut store = Store::memory().unwrap();
        let asset_id = background(&mut store);
        let map = create(&mut store, asset_id, 16, 16, 4, 4).unwrap();
        let project_id = store.asset_read(asset_id).unwrap().project_id;
        let tile_id = tile(&mut store, project_id, 16, 16);

        let placements = vec![Placement {
            x: 0,
            y: 0,
            tile: Some(tile_id.0),
        }];
        let placed = place(&mut store, asset_id, map.layers[0].name.clone(), placements).unwrap();
        assert_eq!(placed.layers[0].tiles[0], Some(tile_id.0));

        let read_back = store.tilemap_read(asset_id).unwrap();
        assert_eq!(read_back, placed);
    }

    #[test]
    fn tiles_lists_same_project_and_size() {
        let mut store = Store::memory().unwrap();
        let asset_id = background(&mut store);
        create(&mut store, asset_id, 16, 16, 4, 4).unwrap();
        let project_id = store.asset_read(asset_id).unwrap().project_id;

        let matching = tile(&mut store, project_id, 16, 16);
        let _wrong_size = tile(&mut store, project_id, 8, 8);
        let other_project = store.project_create("Other", "gameboy").unwrap();
        let _other_project_tile = tile(&mut store, other_project.id, 16, 16);

        let listed = tiles(&mut store, asset_id).unwrap();
        assert_eq!(listed.len(), 1);
        assert_eq!(listed[0].id, matching);
        assert!(listed[0].preview.starts_with("data:image/png;base64,"));
    }

    #[test]
    fn preview_of_one_tile_map_is_data_url() {
        let mut store = Store::memory().unwrap();
        let asset_id = background(&mut store);
        create(&mut store, asset_id, 16, 16, 1, 1).unwrap();

        let url = preview(&mut store, asset_id, None).unwrap();
        assert!(url.starts_with("data:image/png;base64,"));
    }
}
